### PR TITLE
Web and Mac support

### PR DESCRIPTION
**Pull request by saifr383**
In this latest update, I've included support for macOS and web platforms within the Flutter SDK.

### CODE REVIEW

**Comment by malik19995** [comment]
Hello, I've been testing this PR for macOS. 

I've configured an example project. Linked the PR version of the SDK. Facing the following issues on macOS: 

I've added the BundleURL Types entry in Info.plist 

Screenshot 2024-03-27 at 12 10 10 AM

The SDK initializes without an error. But when I trigger sdk.login(), it returns the error below. 


Screenshot 2024-03-27 at 12 09 21 AM

Please let me know if I missed something.  @saifr383

**Comment by malik19995** [comment]
Thank you! 

Added the entitlement. 
Still getting this error. PFA. 

Screenshot 2024-03-27 at 12 30 14 AM

**Comment by saifr383** [comment]
This is mostly due to wrong client id and all

**Comment by saifr383** [comment]
Is this working with starter kit?

**Comment by saifr383** [comment]
> Thank you!
> 
> Added the entitlement. Still getting this error. PFA.
> 
> Screenshot 2024-03-27 at 12 30 14 AM

You are testing it wrong this Pr is only for web support You can test Mac Os support in other
 Pr

**Comment by malik19995** [comment]
Using starter-kit, I face this issue:

No logs are produced.  The app redirects to browser, but never opens anything.

I added the network permission in debug and release entitlements. 

The client ID is as provided by the Kinde portal. The complete env file is copied from there. 

Can you please tell what do you send for loginRedirectUri and logoutRedirectUri in macOS? 

https://github.com/kinde-oss/kinde-flutter-sdk/assets/46687299/fd444de2-65a8-493c-aaf0-3e0ee5da67fd

**Comment by malik19995** [comment]
The issue was with the callback URLs. 

Fixing them, fixed the issue. Thank you for your help 🙌🏼

**Comment by malik19995** [comment]
'flutter test' on the SDK fails. Please add MockBuildContext to the tests.  PFA. 

Screenshot 2024-03-28 at 8 01 02 PM

Screenshot 2024-03-28 at 8 12 18 PM
 
Screenshot 2024-03-28 at 8 12 37 PM

Another issue with the logout test.

**Comment by saifr383** [comment]
> Hello, I've been testing this PR for macOS.
> 
> I've configured an example project. Linked the PR version of the SDK. Facing the following issues on macOS:
> 
> I've added the BundleURL Types entry in Info.plist
> 
> Screenshot 2024-03-27 at 12 10 10 AM
> The SDK initializes without an error. But when I trigger sdk.login(), it returns the error below.
> 
> Screenshot 2024-03-27 at 12 09 21 AM
> Please let me know if I missed something. @saifr383

Updated the code you can test it now

**Comment by malik19995** [comment]
Unable to Create an Organization on the Web, 

PFA error logs. 
Screenshot 2024-04-19 at 2 54 01 PM

**Comment by malik19995** [comment]
https://github.com/kinde-oss/kinde-flutter-sdk/assets/46687299/d907f1fd-da85-4674-af5d-10d1f4875cb3

Facing this issue on web. 

https://kinde-22dbd.web.app/ 

After login, the context does not switch properly. on starterkit

**Comment by saifr383** [comment]
> Screen.Recording.2024-04-21.at.8.18.33.PM.mov 
> Facing this issue on web.
> 
> https://kinde-22dbd.web.app/
> 
> After login, the context does not switch properly. on starterkit

Can you tell the whole steps to replicate it I am unable to replicate it on my end

**Comment by saifr383** [comment]
> Unable to Create an Organization on the Web,
> 
> PFA error logs. Screenshot 2024-04-19 at 2 54 01 PM

error was bcz of context parameter empty has added a exception message for that

**Comment by malik19995** [comment]
Ok perfect. I'll check this out. 

I am also facing another issue on macOS. 

The callback isUserAuthenticated() does not work for macOS. Keeps returning false, even after login is completed and auth token has been obtained.

**Comment by saifr383** [comment]
> Ok perfect. I'll check this out.
> 
> I am also facing another issue on macOS.
> 
> The callback isUserAuthenticated() does not work for macOS. Keeps returning false, even after login is completed and auth token has been obtained.

You have to add 
<key>keychain-access-groups</key>
<array/>

in entitlements as we are using key chain for storing the token

**Comment by malik19995** [comment]
Perfect, works! Thank you! 

Can we please add these setup steps to the README as well?

**Comment by malik19995** [comment]
One more thing,  

You cannot add the `oauth_webauth` package like this. This cannot be merged with the main. Since it needs to be published to pub.dev. PFA.

I suggest either publishing the `oauth_webauth` package and linking via pub.dev or manually adding the package to the repo. 

But it should be added that the local package implementation is linked to the repo so we can pull updates. 

Screenshot 2024-05-03 at 4 54 50 PM

**Comment by saifr383** [comment]
Published the package to Pub Dev and removed the GitHub repository link.

**Comment by malik19995** [comment]
I am unable to run this. I am testing with PR #4 on flutter-starter-kit. 

Please check the screenshot for the error log and Flutter doctor details. 
I believe the ViewConfiguration issue is originating from the oauth_web_plugin. I am on Flutter 3.22.2.

Screenshot 2024-07-12 at 10 54 20 AM

**Comment by saifr383** [comment]
> I am unable to run this. I am testing with PR #4 on flutter-starter-kit.
> 
> Please check the screenshot for the error log and Flutter doctor details. I believe the ViewConfiguration issue is originating from the oauth_web_plugin. I am on Flutter 3.22.2.
> 
> Screenshot 2024-07-12 at 10 54 20 AM

Fixed this issue you can check it by updating the sdk package

File: lib/src/kinde_flutter_sdk.dart [hunk 1]
```diff
@@ -140,51 +148,137 @@ class KindeFlutterSDK with TokenUtils, HandleNetworkMixin {
...
+          params,
+          context!,
+        );
+      } else {
+        if (type == AuthFlowType.pkce) {
+          return _pkceLogin(params);
+        } else {
+          return _normalLogin(params);
+        }
+      }
```
[the remark this answers is not in the document]
Comment (malik19995) [reply lib/src/kinde_flutter_sdk.dart:225]: @saifr383 Please adopt this.

**Comment by malik19995** [comment]
Perfect, this works now! Thank you! @saifr383 

I just added a comment on a code rabbit's review suggestions; please check those out and refactor them.

**Comment by saifr383** [comment]
@malik19995  refactored the code according to code rabbit you can  reveiw it

**Comment by malik19995** [comment]
Cleared for merge 💯

File: lib/src/api/o_auth_api.dart [hunk 2]
```diff
@@ -99,8 +99,22 @@ class OAuthApi {
...
+      method: r'GET',
+    );
+
+    final response = await Dio().request<Object>(
+      path,
+      options: options,
+    );
+
+    return response.statusCode == 200;
+  }
```
[the remark this answers is not in the document]
Comment (AndroidJedi) [reply lib/src/api/o_auth_api.dart]: Web/MacOs test results:

I'm able to signIn and signOut on Web and just SignIn on MacOs.
For signOut It just calls BE API /signout and on Web without any further redirection it just reloads the page:
if (kIsWeb) {
html.window.location.reload();
}

But the same code doesn't work for MacOS, as the compilers are different and for MacOs compiler this is an invalid string interpolation:
const path = r'$url';

but, again for web it works well;

**Review by malik19995 — APPROVED** [review]

**Comment by AndroidJedi** [comment]
test result:
kinde-flutter-sdk works well on Android/iOS/WEB.

on MacOs logout doesn't work, gives error:

```
flutter: ------ OAuthWebAuth appBaseUri: file:///Users/serhiiholovashchuk/Library/Containers/au.kinde.flutterstarterkit.flutterStarterKit1/Data ------
[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: DioException [unknown]: null
Error: Invalid argument(s): No host specified in URI $url
#0 DioMixin.fetch (package:dio/src/dio_mixin.dart:519:7)
<asynchronous suspension>
#1 OAuthApi.logout (package:kinde_flutter_sdk/src/api/o_auth_api.dart:108:22)
<asynchronous suspension>
#2 KindeFlutterSDK.handleWebAndMacLogout (package:kinde_flutter_sdk/src/kinde_flutter_sdk.dart:188:5)
<asynchronous suspension>
#3 KindeFlutterSDK.logout (package:kinde_flutter_sdk/src/kinde_flutter_sdk.dart:167:7)
<asynchronous suspension>
#4 *HomePageState.*signOu
```

**Comment by malik19995** [comment]
The issue on macOS stems from 

`https://pub.dev/packages/flutter_custom_tabs `

This plugin launches the URLs internally within the application via inAppWebView. It is not yet compatible with macOS, so there is no implementation error. 

**We can add a PR to that plugin and request them to merge and release. But this might take some time.** 

I have found an alternative setup that works for us right now. However, the only issue might be an extra redirect to the user's web browser instead. 

We need to add the following case to our SDK logout method. 

PubSpec Update:  (Add Dependency)
`  url_launcher: ^6.3.0
` 

`
else if (Platform.isMacOS) {
      // TODO: REFACTOR:: Upgraded CustomTabs with MacOS Support Plugin is required, to do the callback internally. Right now, it launches in a separate browser window.
      await launchUrl(
        Uri.parse(
          _buildEndSessionUrl().toString(),
        ),
      );
    }
    `
    This uses the default UrlLauncher from the Flutter team. 
    **This opens the URL externally and performs a successful logout.** 
    
Screenshot 2024-08-12 at 6 12 02 AM

**Comment by saifr383** [comment]
> The issue on macOS stems from
> 
> `https://pub.dev/packages/flutter_custom_tabs `
> 
> This plugin launches the URLs internally within the application via inAppWebView. It is not yet compatible with macOS, so there is no implementation error.
> 
> **We can add a PR to that plugin and request them to merge and release. But this might take some time.**
> 
> I have found an alternative setup that works for us right now. However, the only issue might be an extra redirect to the user's web browser instead.
> 
> We need to add the following case to our SDK logout method.
> 
> PubSpec Update: (Add Dependency) ` url_launcher: ^6.3.0`
> 
> `else if (Platform.isMacOS) { // TODO: REFACTOR:: Upgraded CustomTabs with MacOS Support Plugin is required, to do the callback internally. Right now, it launches in a separate browser window. await launchUrl( Uri.parse( _buildEndSessionUrl().toString(), ), ); }` This uses the default UrlLauncher from the Flutter team. **This opens the URL externally and performs a successful logout.**
> 
> Screenshot 2024-08-12 at 6 12 02 AM

We can make a direct GET request instead of opening the same URL in a browser.

**Comment by onderay** [comment]
Closing this PR in favour of the work on this PR - https://github.com/kinde-oss/kinde-flutter-sdk/pull/21

**Pull request closed without merging**